Modify the virtio block to async

liqinggd opened the pull request:
This PR depends on https://github.com/asterinas/asterinas/pull/739

tatetian commented:
@liqinggd This PR can be rebased to the latest master.

@yingdi-shan Could you review this PR?

yingdi-shan commented on kernel/comps/virtio/src/device/block/device.rs:
```diff
@@ -351,28 +394,53 @@ impl DeviceInner {
+struct SubmittedRequest {
+    id: u16,
+    bio_request: BioRequest,
+    bufs: Vec<(DmaStream, usize, usize)>,
```
Maybe we should give a better name or type to the tuple (DmaStream, usize, usize).

liqinggd replied:
In actuality, it is a `DmsStreamSlice`; however, it needs to maintain ownership of `DmaStream` (this is because `DmaStream` does not retain ownership but is merely a reference).

yingdi-shan replied:
Maybe we should consider implementing an owned type for `DmaStreamSlice` ?

liqinggd replied:
I think using a tuple is simple enough to represent the `DmaStream`, and we can use it to construct the `DmaStreamSlice`

yingdi-shan replied:
Okay

yingdi-shan commented on kernel/comps/virtio/src/device/block/device.rs:
```diff
@@ -279,17 +244,72 @@ impl DeviceInner {
+
+    /// Reads data from the device, this function is no-blocking.
+    fn do_read(&self, bio_request: BioRequest) {
+        let dma_streams: Vec<(DmaStream, usize, usize)> = bio_request
+            .bios()
+            .flat_map(|bio| {
+                bio.segments().iter().map(|segment| {
+                    let dma_stream =
+                        DmaStream::map(segment.pages().clone(), DmaDirection::FromDevice, false)
+                            .unwrap();
+                    (dma_stream, segment.offset(), segment.nbytes())
+                })
+            })
+            .collect();
+        let dma_slices: Vec<DmaStreamSlice> = dma_streams
+            .iter()
+            .map(|(stream, offset, len)| DmaStreamSlice::new(stream, *offset, *len))
+            .collect();
```
The `do_write` function contains duplicated code segments. Refactoring this code into a standalone function would enhance readability and maintainability.

liqinggd replied:
Maybe  I can add a function to do the dma mapping

liqinggd replied:
Add `dma_stream_map`.

yingdi-shan reviewed (commented):
This PR modifies the virtio block to async. I think this PR is critical for further improving the performance of our storage stack.

liqinggd commented on kernel/comps/virtio/src/device/block/device.rs:
```diff
@@ -148,76 +92,95 @@ struct DeviceInner {
             .unwrap();
 
         device
             .transport
-            .register_queue_callback(0, Box::new(handle_block_device), false)
+            .register_queue_callback(0, Box::new(dummy_handle_block_device), false)
             .unwrap();
```
The reason the interruption stopped working before is because in the previous implementation, there was no callback function set when calling `get_id`. This led to the inability to receive interruptions later on, even after setting a callback function.

yingdi-shan replied:
The root of our issue appears to be the need to obtain the `device_id` for locating the `block_device`, which is essential for handling block interrupts. However, this step might be unnecessary. By passing the `block_device` pointer directly into the init function, we could defer retrieving the `device_id` until after all other initialization functions have completed. This approach could streamline our initialization process.

liqinggd replied:
I moved the code into `init` function, and add some comments here to explain why we set the callback function twice

yingdi-shan replied:
Thanks~

liqinggd commented on kernel/comps/virtio/src/device/block/device.rs:
```diff
@@ -148,76 +92,95 @@ struct DeviceInner {
         Ok(device)
     }
 
+    /// Finalizes the device initialization and assigns the device ID.
+    pub fn finish_init(&mut self, device_id: String) {
+        assert!(self.device_id.is_none());
+
+        let cloned_device_id = device_id.clone();
+        self.device_id = Some(cloned_device_id);
+
+        let handle_block_device = move |_: &TrapFrame| {
+            aster_block::get_device(device_id.as_str())
+                .unwrap()
+                .handle_irq();
+        };
+        self.transport
+            .register_queue_callback(0, Box::new(handle_block_device), false)
+            .unwrap();
+
+        self.transport.finish_init();
+    }
```
This logic actually follows the design of Transport, but indeed, it increases the likelihood of using the API incorrectly.

tatetian commented on kernel/aster-nix/src/lib.rs:
```diff
@@ -80,6 +80,7 @@ fn init_thread() {
         "[kernel] Spawn init thread, tid = {}",
         current_thread!().tid()
     );
+    aster_frame::enable_interrupts();
```
I don't understand. Is this single-line change a bugfix? When are the local IRQs enabled before this change?

liqinggd replied:
> I don't understand. Is this single-line change a bugfix?

We must enable interrupts before loading the fs, I think If we use `mount` syscall to do the mounting in the init process. this API can be removed. I will add some comments here to expain it.

> When are the local IRQs enabled before this change?

I searched the code and I found it is enabled in the `execute` function of User Context.

tatetian replied:
> I searched the code and I found it is enabled in the execute function of User Context.

Ok. Then, this is a bugfix. Interrupts should be enabled far earlier. I will accept this line of change as a quick fix. But a permanent solution should be discussed and fixed in a separate PR. Could you add a FIXME comment to this line?

liqinggd replied:
Done

tatetian commented:
@yingdi-shan Any more comments for the updated version?

yingdi-shan commented:
I am satisfied with the updated version. @tatetian

tatetian commented on kernel/comps/virtio/src/device/block/device.rs:
```diff
@@ -186,31 +185,40 @@ impl DeviceInner {
+
+        // Sets the queue callback function properly before calling `get_id`;
+        // otherwise, the interrupt mechanism cannot work.
         let device_id = device.get_id();
```
The verb get is usually reserved for light-weight methods. This `get_id` method involves I/O and blocking. So it should be considered heavy-weight. I suggest renaming it to `fetch_id` or `request_device_id`.

liqinggd replied:
Done

tatetian commented on kernel/comps/virtio/src/device/block/device.rs:
```diff
@@ -186,31 +185,40 @@ impl DeviceInner {
-
         device
             .transport
             .register_queue_callback(0, Box::new(handle_block_device), false)
```
The device ID is required before registering this callback because the `handle_block_device` uses the ID internally.

```rust
        let handle_block_device = move |_: &TrapFrame| {
            aster_block::get_device(device_id.as_str())
                .unwrap()
                .handle_irq();
        };
```

But the `handle_irq` method is implemented by the virtio-block driver.

```rust
impl aster_block::BlockDevice for BlockDevice {
    fn enqueue(&self, bio: SubmittedBio) -> Result<(), BioEnqueueError> {
        self.queue.enqueue(bio)
    }

    fn handle_irq(&self) {
        info!("Virtio block device handle irq");
        self.device.do_handle_irq();
    }
}
```

In other words, the driver of a block device---in this case, the virtio-block driver---has all the needed information to process the IRQ. It does not need to rely on the block component to process the IRQ. As such, requesting the device ID before registering the IRQ callback is unnecessary. I think this issue has been identified by @yingdi-shan. And the comment is marked resolved. But I don't think this is resolved.

A simple fix is shown below.

```rust
impl DeviceInner {
    // The return type is modified to return `Arc<Self>` on success.
    fn init(mut transport: Box<dyn VirtioTransport>) -> Result<Arc<Self>, VirtioDeviceError> {
        // ...

        let mut device = Arc::new(Self {
            // ...
        });

        // === Removed ===
        // device
        //    .transport
        //    .register_queue_callback(0, Box::new(dummy_handle_block_device), false)
        //    .unwrap();
        // let device_id = device.get_id();
        // === Removed ===

        let handle_block_device = {
            let cloned = device.clone();
            move |_: &TrapFrame| {
                cloned.handle_irq();
            };
        };
        device
            .transport
            .register_queue_callback(0, Box::new(handle_block_device), false)
            .unwrap();

        Ok(device)
    }
}

```

The new version simplifies the `init` method. More importantly, it avoids looking the block device table every time a IRQ is fired. So the new version also improve efficiency.

liqinggd replied:
This gives me a good hint; in fact, the device id does not need to be stored in the device information. Following this idea, I have rewritten this part of the logic.

tatetian commented on framework/aster-frame/src/lib.rs:
```diff
@@ -69,6 +69,16 @@ pub fn init() {
+/// it is necessary to have this API to enable interrupts for the proper functioning of block devices.
+/// If we move the step of mounting the filesystems to be done within the init process,
+/// this API can be removed.
+pub fn enable_interrupts() {
```
First of all, this function should not be added the root module of `aster-frame`. There is the `trap` module, which is most related to IRQ. Move the function to this module. In fact, you can simple make the existing `arch::irq::enable_local()` function public. The proposed `enable_interrupts` and the existing `enable_local` are basically the same thing.

Second, the soundness of this function is questionable as exposing this function allows the Framework's users to enable interrupts when she should not. For example, she may write code.

```rust
let guard = lock.lock_irq_disabled();
enable_interrupts();
```

To say the least, exposing `enable_interrupts` breaks the assumption of `SpinLock::lock_irq_disabled()`. So it should be avoided.

So I suggest ading a FIXME.

```rust
// FIXME: this method may be unsound.
pub fn enable_interrupts() {
```

Third, the Rust doc seems to miss the point.

```rust
/// Currently, since the filesystems are mounted during the OS initialization phase,
/// it is necessary to have this API to enable interrupts for the proper functioning of block devices.
/// If we move the step of mounting the filesystems to be done within the init process,
/// this API can be removed.
```

The point is that _IRQs should be enabled as early as possible_. The quick fix in this PR, including adding this `enable_interrupts` method, suggests that IRQs are enabled as late as the time the CPU enters the user space for the first time. This is too late.

liqinggd replied:
I am not quite familiar with the code structure of `trap`, so I just add an API in lib.rs. 

I have moved the code into the irq.rs of trap module, and refine the expalin of this API.

liqinggd commented:
After I rebased the PR, the CI tests often encounter an issue where they hang during `/regression/ext2.sh`. Firstly, the system bootup with `SCHEME=microvm` itself often triggers a panic, indicating that it is unstable. As a result, I modified the CI script to run the regression test in `BOOT_PROTOCOL=multiboot2` mode, which is more stable.

Furthermore, since this hanging issue occurs intermittently, I refined the script to allow it to run multiple times.

Lastly, I discovered that the issue of hanging is more likely to occur during the evicting of multiple blocks at once in ext2 evict block. As a workaround, I temporarily set this number to 1.

tatetian commented:
> the system bootup with `SCHEME=microvm` itself often triggers a panic

If this is a bug, file an issue.

> Furthermore, since this hanging issue occurs occasionally, I refined the script to allow it to run multiple times.
>
> Lastly, I discovered that the issue of hanging is more likely to occur during the evicting of multiple blocks at once in ext2 evict block. As a workaround, I temporarily set this number to 1.

Introducing workarounds should be our last resort. Could you try harder to identify the root cause of the hanging bug?

liqinggd commented:
> SCHEME=microvm



> > the system bootup with `SCHEME=microvm` itself often triggers a panic
> 
> If this is a bug, file an issue.

OK, I will report a bug for it.
> 
> > Furthermore, since this hanging issue occurs occasionally, I refined the script to allow it to run multiple times.
> > Lastly, I discovered that the issue of hanging is more likely to occur during the evicting of multiple blocks at once in ext2 evict block. As a workaround, I temporarily set this number to 1.
> 
> Introducing workarounds should be our last resort. Could you try harder to identify the root cause of the hanging bug?

It is hard to obtain the problematic scene when hanging. I attempt to use a GDB server, but upon doing so, the issue no longer arises. This problem arises only when the program runs neither too fast nor too slow. Adding logs to slow down the program also causes the issue to vanish. It currently appears to be a race condition problem.

tatetian commented:
Regarding the hanging problem, I think I have some clues.

The fundamental change introduced in this PR is to add a real IRQ handler for the virtio-block driver. Previously, this IRQ handler is a [no-op](https://github.com/asterinas/asterinas/blob/698e7481508712a477fd8f043fc25951051324a3/kernel/comps/virtio/src/device/block/device.rs#L139).

But the author of an IRQ handler must be very careful: the IRQ part of the driver must NOT share a `SpinLock` with the normal part of the driver unless the `SpinLock` is acquired with IRQ disabled in the task context. Failing to follow the this rule leads to deadlock.

The latest version of this PR, unfortunately, does NOT follow this rule. For example, `submitted_requests` is protected with `SpinLock` and used in both the task and interrupt contexts. But the lock on `submitted_requests` is not acquired with IRQ disabled. The same problem happens on `id_allocator`.

tatetian reviewed (approved):
LGTM. Thanks for the contribution!
